Draft: Provide initialization documentation for cloning Dandi ecosystem (#104)
dandi/handbook: docs/60_initialize_vendors.md

# Initialize Vendor Accounts

The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:

> - **Heroku**: Provisions the API components

Reviewer: ```suggestion
- **Heroku**: Provisions the API components.
```

for consistency

> - **AWS**: Provides storage buckets, as well as domain management, for resources across the DANDI ecosystem. As well as the services (Kubernetes, etc.) for deploying the JupyterHub.

Reviewer: ```suggestion
- **AWS**: Provides storage buckets (S3), as well as domain management (Route53), for resources across the DANDI ecosystem. As well as the services (EC2, Kubernetes, etc.) for deploying the JupyterHub.
```
?

You'll need to create a GitHub Organization where you can fork the DANDI repositories. [See here for documentation to create a GitHub organization](https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/creating-a-new-organization-from-scratch)

##### Initialize your OAuth App

Once you create the Organization, navigate to the `Settings` tab:

<br/><br/>
<img
src="../img/github_home.jpg"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Under `Settings`, you'll want to initialize an `OAuth App` -- navigate to `Developer settings > OAuth Apps`

<br/><br/>
<img
src="../img/oauth_app.jpg"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Click on `New Org OAuth App` next

<br/><br/>
<img
src="../img/new_app.jpg"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You'll be prompted with the following form -- see the example values populated in this screenshot -- more to come in other sections for where these values might be populated:

<br/><br/>

> <img

Reviewer 2: We (EMBER) ended up needing to change these to the following to get auth working:

- Homepage URL: https://\<your-awesome-subdomain-if-applicable>.\<your-awesome-domain>.org
   - https://dandi.emberarchive.org
- Authorization callback URL:  https://\<your-awesome-domain>.org
   - https://emberarchive.org 

rather than the existing:
![image](https://github.com/user-attachments/assets/02878a27-58aa-4a45-867e-53fd68958919)

# Initialize Vendor Accounts

The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:

> • **Heroku**: Provisions the API componenents

Reviewer 3: ```suggestion
• **Heroku**: Provisions the API components
```

Author: Updated!

• **AWS**: Provides storage buckets, as well as domain management, for resources across the DANDI ecosystem. As well as the services (Kubernetes, etc.) for deploying the JupyterHub.
• **GitHub**: Serves as the authentication provider for accounts across the DANDI ecosystem
• **Terraform Cloud**: Manages provisioned resources across cloud vendors in a version-controlled manner. 
• **Netlify**: Deploys production frontend build, as well as staging previews to assist with frontend development

> • **Sentry**: Provides observability and monitoring for API events

Reviewer 3: Not sure why, but the bullet points don't properly render as a list.

![image](https://github.com/user-attachments/assets/9ea81306-f538-428b-8bed-a6f4b93b3c37)

Author: Updated!

Some services are not yet integrated within the main infrastructure:

• https://github.com/dandi/backups2datalad - to populate/update https://github.com/dandi/dandisets, https://github.com/dandisets, and https://github.com/dandizarrs/

• TODO: heroku logs

• TODO: aws s3 access stats dump

• TODO: con/tinuous dumps of CI logs 

• TODO: zarr manifests generation (ATM not on drogon even)

> • TODO: access stats analysis/plots (yet to be finished/cron deployed)

Reviewer 4: should these remain as TODOs?

Reviewer 3: I have filed issues in this repo.  These TODOs can now be removed from this doc.

> • TODO: heroku logs
> 
> • TODO: aws s3 access stats dump
> 
> • TODO: con/tinuous dumps of CI logs 
> 
> • TODO: zarr manifests generation (ATM not on drogon even)
> 
> • TODO: access stats analysis/plots (yet to be finished/cron deployed)

Reviewer 3: ```suggestion
```

Author: Updated!

##### Initialize your OAuth App

Once you create the Organization, navigate to the `Settings` tab:

<br/><br/>
<img
src="../img/github_home.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Under `Settings`, you'll want to initialize an `OAuth App` -- navigate to `Developer settings > OAuth Apps`

<br/><br/>
<img
src="../img/oauth_app.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Click on `New Org OAuth App` next

<br/><br/>
<img
src="../img/new_app.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You'll be prompted with the following form -- see the example values populated in this screenshot -- more to come in other sections for where these values might be populated:

<br/><br/>
<img
src="../img/register_new.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Obtaining your OAuth App Credentials

> After creating your OAuth App, you'll lastly want to obtain a client secret key and your client ID -- make sure to note these values for further steps when creating our API

Reviewer 3: ```suggestion
After creating your OAuth App, you'll lastly want to obtain a client secret key and your client ID -- make sure to note these values for further steps when creating our API.
```

Author: Updated!

<br/><br/>

##### Connecting GitHub with Heroku

**This step is contingent that you have forked [dandi-archive](https://github.com/dandi/dandi-archive)**

Once you have the repository forked, you'll want to navigate to `Settings` in the repository page:

<br/><br/>
<img
src="../img/linc_settings.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Next, navigate to `Security > Secrets and variables > Actions`

<br/><br/>
<img
src="../img/secrets_actions.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

And then lastly create two new secrets, `HEROKU_EMAIL` should be the email address/account you used to generate your `HEROKU_API_KEY`

<br/><br/>
<img
src="../img/heroku_secrets.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

## Terraform Cloud

Terraform is configuration tool for managing "infrastructure-as-code" -- meaning that we can programmatically manage infrastructure in
a traceable, version-controlled form.

> The Terraform ecosystem provides a UI tool called **Terraform Cloud**

Author: Updated!

<br/><br/>

You'll want to next connect `dandi-archive` via `Deploy with GitHub`. Enable Netlify to be authorized as a GitHub app. Once you have
enabled authorization, you'll need to specifically link your appropriate repositories:

<br/><br/>
<img
src="../img/netlify_auth.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Now once you see the appropriate repository, you'll want to navigate to configure where Netlify can find and build your site

<br/><br/>
<img
src="../img/netlify_repo_search.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

For usage of Netlify, one could refer to declaring a `netlify.toml` configuration file [like the one referenced in DANDI Archive](https://github.com/dandi/dandi-archive/blob/master/web/netlify.toml)

These values can also be replicated in the settings. 

<br/><br/>
<img
src="../img/netlify_config.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> Your frontend should be able to deploy to an auto-generated URL via Netlify now! Steps for domain management and configuration are described further in the [Frontend Deployment](../64_dandi_archive/#frontend-deployment) section of these within the DANDI Archive setup.

Reviewer 3: ```suggestion
Your frontend should be able to deploy to an auto-generated URL via Netlify now! Steps for domain management and configuration are described further in the [Frontend Deployment](../64_dandi_archive/#frontend-deployment) section within the DANDI Archive setup.
```

Author: Updated!

## Sentry

> [Sentry](https://sentry.io/) is a monitoring tool used for the DANDI Archive API. It is integral in order to notify engineers if a system is down, experiencing poor performance, or may have unwanted users

Reviewer 3: ```suggestion
[Sentry](https://sentry.io/) is a monitoring tool used for the DANDI Archive API. It is integral in order to notify engineers if a system is down, experiencing poor performance, or may have unwanted users.
```

Author: Updated!

Begin by creating a Sentry account -- once successful, you'll start by creating a new Project:

<br/><br/>
<img
src="../img/sentry_init.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> For your account, the free `Developer` account should be accurate, as define in [Sentry's pricing structure](https://sentry.io/pricing/)

Reviewer 3: ```suggestion
For your account, the free `Developer` account should be sufficient, as defined in [Sentry's pricing structure](https://sentry.io/pricing/).
```

Author: Updated!

##### Select Django as an App Type

DANDI Archive API is built as a Django app -- so proceed to select `Django` on the following screen:

<br/><br/>
<img
src="../img/sentry_django.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Capture Sentry DSN value

You'll be provided with a screen displaying how to initialize and install Sentry into your Django app. For now, just capture the DSN value.
This value will be used later as an environment variable while deploying your API via Terraform.

<br/><br/>
<img
src="../img/sentry_sdk_dsn.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> ## datalad (TBD)
> 
> ## git-annex (TBD)

Reviewer 3: ```suggestion
## datalad

We will add docs in the future (https://github.com/dandi/dandi-docs/issues/132).

## git-annex

We will add docs in the future (https://github.com/dandi/dandi-docs/issues/132).
```

# Initialize Vendor Accounts

The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:

• **Heroku**

Provisions the servers and worker processes for the API. Heroku also handles the Postgres instance responsible
for data models in the DANDI Archive.

• **AWS**

> Provides storage buckets, as well as domain management, for resources across the DANDI ecosystem 

Reviewer 3: ```suggestion
Provides storage buckets, as well as domain management, for resources across the DANDI ecosystem.  As well as the services (Kubernetes, etc.) for deploying the JupyterHub.
```

Author: Updated!

> # Initialize Vendor Accounts

Author: @waxlamp @yarikoptic @satra @kabilar would be great if you could review

Reviewer 5: Needs Dockerhub (or some other container registry)

Reviewer 3: Thanks, yes, we should add a section on Docker Hub.

Reviewer 3: After discussion with Aaron, we didn't need to create a Docker Hub account for LINC since we are using the DANDI JupyterHub images.  (We did create one for our Webknossos images but that is out of scope from this pull request.)

The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:

• **Heroku**

• **AWS**

• **GitHub**

• **Terraform Cloud**

• **Netlify**

• **Sentry**

> • **PyPI**

Reviewer 4: i think it would be good to add what each of these services provide to the infrastructure.

Author: Thanks Satra -- included a brief blurb for what each service is responsible for

> • **Datalad (TBD)**
> 
> • **git-annex (TBD)**

Reviewer: Those are just tools -- no account needed and overall they just rely on above services (GitHub) and git configuration to run.  

But for completeness -- we do need a host (in DANDI case it is `drogon` server) with an account under which to run all those additional "cron jobs", hence

```suggestion

In addition a host (local server or an instance in cloud) is needed to run additional services employing DataLad and git-annex.
```

Reviewer 4: are cron-jobs part of the infrastructure? or should be considered essential?

Reviewer: We do not have definition of "infrastructure" to define the boundary. 

if we consider https://github.com/dandisets etc as part of infrastructure, then yes.

as "essential" -- likely not as long as not integrated within dandiarchive.org web UI.

<br/><br/>
<img
src="../img/sentry_sdk_dsn.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

## PyPI

A common way that users interact with the DANDI ecosystem is via its CLI and Python client tool: [dandi-cli](https://pypi.org/project/dandi/)

##### Create a PyPI Account

Versions of this tool are hosted, versioned and managed by PyPI -- therefore, you will need to create an account.

##### Retrieve an API Token

After creating an account, you'll want to create your own API key -- first go to `Account settings`

<br/><br/>
<img
src="../img/pypi_account.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Scroll down to `API tokens` and create a value -- keep track of this value for now. We will use it later

<br/><br/>
<img
src="../img/pypi_api_token.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

In terms of creating and publishing your first project, go to the [documentation for setting up a CLI and Python Client](.
./47_cli_python)

> ## datalad (TBD)
> 
> ## git-annex (TBD)

Reviewer: @yarikoptic needs to finish up but likely elsewhere and here just describe setup of the box:

```suggestion
## A host for extra services

Some services are not yet integrated within the main infrastructure:

- https://github.com/dandi/backups2datalad - to populate/update https://github.com/dandi/dandisets, https://github.com/dandisets, and https://github.com/dandizarrs/
- TODO: heroku logs
- TODO: aws s3 access stats dump
- TODO: con/tinuous dumps of CI logs
- TODO: zarr manifests generation (ATM not on drogon even)
- TODO: access stats analysis/plots (yet to be finished/cron deployed)
```

Author: Thanks Yarik -- updated

# Initialize Vendor Accounts

This series of documents explains how to create a new instance of the DANDI project (i.e. a clone of the entire DANDI ecosystem). It is suggested that you briefly read through each of the
documents in this guide before starting.

The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:

• **Heroku**

• **AWS**

• **GitHub**

• **Terraform Cloud**

• **Netlify**

• **Sentry**

• **PyPI**

• **Datalad (TBD)**

• **git-annex (TBD)**

## Heroku

##### Create your own Heroku account

No special steps here, just create!

##### Create a "Team" 

Send invites to the appropriate e-mail addresses to onboard your team

##### Create an "App"

<br/><br/>
<img

> src="../img/heroku_create_app.png"

Reviewer 3: I believe the `img` directory is within `docs`.  Will need to change the path for all images.

```suggestion
src="./img/heroku_create_app.png"
```

Author: Hmmm changing this broke all the image references -- I think it is working as intended (unless you are observing something different locally?)

Reviewer 3: Thanks for checking.  The references are broken in the GitHub UI, but I didn't test locally with MkDocs.  Perhaps the working directory for MkDocs is different so it works out.  Let's ignore this suggestion.

##### Create an "App"

<br/><br/>
<img
src="../img/heroku_create_app.png"
alt="heroku_create_app"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You can name whatever you wish -- no need to add a pipeline, etc. **Be sure to note the app name**

<br/><br/>
<img
src="../img/heroku_app_name.png"
alt="heroku_app_name"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Your app will be an empty template, but that is OK -- more to come here!

##### Obtain Heroku API Key

You'll need to provide access to Heroku for GitHub, Terraform, etc. -- thus, you'll need to generate an API Key

First go to your `Account Settings`:

<br/><br/>
<img
src="../img/heroku_settings.png"
alt="heroku_app_name"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Next, find the `API Key` section and generate a new key.

<br/><br/>
<img
src="../img/heroku_key.png"
alt="heroku_app_name"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> Keep this value for further steps.

Reviewer: It would be valuable to know what sizes of instances we have for DANDI and LINC to guide installations of other instances.

Author: Sure -- I can add, they are defined in the DANDI Infra `api.tf` Girder extension here: https://github.com/dandi/dandi-infrastructure/blob/master/terraform/api.tf#L14-L18

On this note, has any stress-testing ever been done to evaluate if these worker sizes are appropriate or not for DANDI?

Reviewer 4: we have bumped on occasion in the early days when receiving http errors. but no serious stress testing was applied. i would say that independent of workshops, conferences, or class settings, we have not run into significant sustained lags with the setup. some of this is now monitored by upptime.

<br/><br/>

You'll want to next connect `dandi-archive` via `Deploy with GitHub`. Enable Netlify to be authorized as a GitHub app. Once you have
enabled authorization, you'll need to specifically link your appropriate repositories:

<br/><br/>
<img
src="../img/netlify_auth.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Now once you see the appropriate repository, you'll want to navigate to configure where Netlify can find and build your site

<br/><br/>
<img
src="../img/netlify_repo_search.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

For usage of Netlify, one could refer to declaring a `netlify.toml` configuration file [like the one referenced in DANDI Archive](https://github.com/dandi/dandi-archive/blob/master/web/netlify.toml)

These values can also be replicated in the settings. 

<br/><br/>
<img
src="../img/netlify_config.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> Your frontend should be able to deploy to an auto-generated URL via Netlify now! Steps for domain management and configuration are described further in the [Frontend Deployment](../64_dandi_archive/#frontend-deployment) section of these within the DANDI Archive setup.

Reviewer: Question came up on knowing how many minutes is needed by netlify for DANDI instance.

Author: When you say `minutes` in this context for Netlify, do you mean "build minutes"? (e.g. how long Netlify runners are required to run to deploy?) Or something else?

• TODO: heroku logs

• TODO: aws s3 access stats dump

• TODO: con/tinuous dumps of CI logs 

• TODO: zarr manifests generation (ATM not on drogon even)

• TODO: access stats analysis/plots (yet to be finished/cron deployed)

## Heroku

##### Create your own Heroku account

No special steps here, just create!

##### Create a "Team" 

Send invites to the appropriate e-mail addresses to onboard your team

##### Create an "App"

<br/><br/>
<img
src="../img/heroku_create_app.png"
alt="heroku_create_app"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You can name whatever you wish -- no need to add a pipeline, etc. **Be sure to note the app name**

<br/><br/>
<img
src="../img/heroku_app_name.png"
alt="heroku_app_name"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Your app will be an empty template, but that is OK -- more to come here!

##### Obtain Heroku API Key

You'll need to provide access to Heroku for GitHub, Terraform, etc. -- thus, you'll need to generate an API Key

First go to your `Account Settings`:

<br/><br/>

> <img

Reviewer 2: This image was a bit mis-leading when following the instructions. The image led us to believe we needed to go to our Heroku App -> Settings, however what we ended up needing to do was the following:
- Click the Profile icon in the top right
- Select "Account Settings" in the dropdown
- Scroll to find the "API Key" section
